### PR TITLE
Adding Guest Agent based logging to VmSnapshotLinux extension

### PR DESCRIPTION
**Pull request by arisettisanjana**
Currently only backup initiated extension logs (of only last successful retry of one run) are sent to Kusto -TraceLogMessages. RestorePoint initiated extension job's logs are not available in Kusto. To obtain logs of each extension run, we utilize the event logging feature provided by GuestAgent, push extension telemetry to AzCore Kusto cluster's fa database.

Changes made:
Modified handle.py to include Logger initialization.
Updated HandlerUtil.py to capture event_folder location and get the severity of logs to be pushed to Kusto
Added the generation and handling of events in EventLoggerUtil.py
Event.py specifies the Schema required by GuestAgent for pushing the logs

### CODE REVIEW

**Review by iamashwani-who — COMMENTED** [review]
validate against different versions of python.

File: VMBackup/main/Utils/Event.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from datetime import datetime
+import os
+import threading
+
+class Event:
+    '''
+         The agent will only pick the first 3K - 3072 characters.
+         Rest of the characters would be discarded from the messages
```
[the remark this answers is not in the document]
Comment (arisettisanjana) [reply VMBackup/main/Utils/Event.py:8]: So now in our logic if message has more than 3k-3072 characters then we divide the message accordingly.
But if this is not the case(i.e our message has more than 3000 characters ) then guest agent would be only pushing first 3k characters and rest would be discarded.
As we are handling this case, the rest would also be pushed.

[the remark this answers is not in the document]
Comment (iamashwani-who) [reply VMBackup/main/Utils/Event.py:12]: OK. can you please change the description to describe the same.

File: VMBackup/main/Utils/Event.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+from datetime import datetime
+import os
+import sys
+if sys.version_info[0] == 3:
+    import threading
+else:
+    import thread as threading
```
Comment (iamashwani-who) [inline VMBackup/main/Utils/Event.py:7]: add a comment in else block to indicate that it is compatibility with python2.

File: VMBackup/main/Utils/EventLoggerUtil.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
+import os
+import threading
+import json
+import sys
+import datetime
+import time
+if sys.version_info[0] == 2:
+    import Queue as queue
+else:
+    import queue
```
Comment (iamashwani-who) [inline VMBackup/main/Utils/EventLoggerUtil.py:10]: add comments to describe the reason for it.

File: VMBackup/main/Utils/EventLoggerUtil.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
+import os
+import threading
```
Comment (iamashwani-who) [inline VMBackup/main/Utils/EventLoggerUtil.py:2]: do we not need compatibility check for it like we needed in Event.py

File: VMBackup/main/Utils/EventLoggerUtil.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+            
+            self.event_processing_interval = LoggingConstants.MinEventProcesingInterval
+            print("Information: Setting event reporting interval to %ss" %(self.event_processing_interval))
+            
+            self.begin_event_queue_polling()
+        else:
+            print("Warning: EventsFolder parameter is empty. Guest Agent does not support event logging.")
+            
+    @staticmethod
+    def GetInstance(log, event_directory, severity_level):
```
Comment (iamashwani-who) [inline VMBackup/main/Utils/EventLoggerUtil.py:64]: q: what conditions could lead to race condition that we need locking in the first place? If this instance is crated in handle.py, would that be enough?

File: VMBackup/main/Utils/HandlerUtil.py [hunk 6]
```diff
@@ -780,6 +797,20 @@ def command_output_from_subprocess(self , args, process_wait_time):
         out = str(out)
         return out
 
+    def get_severity_level(self):
+        logging_level = LoggingLevel(LoggingConstants.DefaultEventLogLevel)
```
Comment (iamashwani-who) [inline VMBackup/main/Utils/HandlerUtil.py:801]: Is the default level set for 2? If not, do we need to create a separate constant for guest agent Kusto Logs?

File: VMBackup/main/handle.py [hunk 7]
```diff
@@ -69,12 +70,16 @@ def main():
...
         total_used_size = 0
         size_calculation_failed = False
+        eventlogger = None
         HandlerUtil.waagent.LoggerInit('/dev/console','/dev/stdout')
         hutil = HandlerUtil.HandlerUtility(HandlerUtil.waagent.Log, HandlerUtil.waagent.Error, CommonVariables.extension_name)
         backup_logger = Backuplogger(hutil)
         MyPatching, patch_class_name, orig_distro = GetMyPatching(backup_logger)
         hutil.patching = MyPatching
         configSeqNo = -1
+        hutil.try_parse_context(configSeqNo)
```
Comment (iamashwani-who) [inline VMBackup/main/handle.py:80]: this looks unrelated to logging change. why is this required?

File: VMBackup/main/Utils/EventLoggerUtil.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+
+            space_available_in_event_directory -= file_size
+        else:
+            space_available_in_event_directory = 0
+            FileHelpers.deleteFile(file_path)
+            print("Information: Event reporting has paused due to reaching maximum capacity in the Event directory. Reporting will resume once space is available. Events for this iteration will not be reported.")
+
+    def dispose(self):
+        print("Information: Dispose(), called on EventLogger. Event processing is terminating...")
+        self._dispose(True)
```
Comment (iamashwani-who) [inline VMBackup/main/Utils/EventLoggerUtil.py:233]: Is the dispose pattern required and applicable?

File: VMBackup/main/Utils/EventLoggerUtil.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+                if EventLogger._instance is None:
+
+                    EventLogger._instance = EventLogger(log, event_directory, severity_level)
+        return EventLogger._instance
+        
+    def update_properties(self, task_id):
+        self.operation_id = task_id
+
+    def trace_message(self, severity_level, message, *args):
+        level = 0
```
Comment (ankitha-p) [inline VMBackup/main/Utils/EventLoggerUtil.py:76]: default severity level should be 2.

File: VMBackup/main/Utils/EventLoggerUtil.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+        if (severity_level == "Verbose"):
+            level = 0
+        elif (severity_level == "Info"):
+            level = 1
+        elif (severity_level == "Warning"):
+            level = 2
+        elif(severity_level == "Error"):
+            level = 3
```
Comment (ankitha-p) [inline VMBackup/main/Utils/EventLoggerUtil.py:84]: Let's separate this section of code, into an enum outside of trace_message fn logic.

File: VMBackup/main/handle.py [hunk 11]
```diff
@@ -317,6 +322,10 @@ def daemon():
         para_parser = ParameterParser(protected_settings, public_settings, backup_logger)
         hutil.update_settings_file()
 
+        if(para_parser.taskId is not None and para_parser.taskId != ""):
+            eventlogger.update_properties(para_parser.taskId)
+        hutil.set_event_logger(eventlogger)
```
Comment (ankitha-p) [inline VMBackup/main/handle.py:327]: why is L:327 required when we are already setting the event logger in L82?

File: VMBackup/main/handle.py [hunk 12]
```diff
@@ -603,20 +614,22 @@ def enable():
         para_parser = ParameterParser(protected_settings, public_settings, backup_logger)
 
         if(para_parser.taskId is not None and para_parser.taskId != ""):
+            eventlogger.update_properties(para_parser.taskId)
+            hutil.set_event_logger(eventlogger)
```
Comment (ankitha-p) [inline VMBackup/main/handle.py:618]: this section of code is duplicate in enable() and daemon(). If we have only one instance of the event logger and taskId doesn't change in one execution. setting it once would suffice.

File: VMBackup/main/handle.py [hunk 13]
```diff
@@ -603,20 +614,22 @@ def enable():
...
-            taskIdentity.save_identity(para_parser.taskId)
-
+            taskIdentity.save_identity(para_parser.taskId)       
         temp_status= 'success'
         temp_result=CommonVariables.ExtensionTempTerminalState
         temp_msg='Transitioning state in extension'
         blob_report_msg, file_report_msg = get_status_to_report(temp_status, temp_result, temp_msg, None)
 
         status_report_to_file(file_report_msg)
+        eventlogger.dispose()
```
Comment (ankitha-p) [inline VMBackup/main/handle.py:632]: what is the idea behind not having one event logger (both for daemon and enable)? Aren't we following singleton pattern?

File: VMBackup/main/Utils/HandlerUtil.py [hunk 14]
```diff
@@ -78,6 +79,8 @@
 import subprocess
 import datetime
 import Utils.Status
+from Utils.EventLoggerUtil import EventLogger
+from Utils.LogHelper import LoggingLevel,LoggingConstants
```
Comment (ankitha-p) [inline VMBackup/main/Utils/HandlerUtil.py:83]: nit: use space after comma

File: VMBackup/main/Utils/EventLoggerUtil.py (hunk 4 above)
Comment (arisettisanjana) [reply VMBackup/main/Utils/EventLoggerUtil.py:2]: so threading module is present in both the versions, just that getting the id of thread is not compatible.

File: VMBackup/main/Utils/HandlerUtil.py [hunk 15]
```diff
@@ -154,6 +164,7 @@ def log(self, message,level='Info'):
         except Exception as e:
             try:
                 errMsg='Exception in hutil.log'
+                self.log_with_no_try_except(e, 'Warning')
                 self.log_with_no_try_except(errMsg, 'Warning')
```
Comment (ankitha-p) [inline VMBackup/main/Utils/HandlerUtil.py:168]: If we are logging exception. L168 can be skipped if we append the message to L167.

File: VMBackup/main/Utils/HandlerUtil.py (hunk 6 above)
Comment (arisettisanjana) [reply VMBackup/main/Utils/HandlerUtil.py:801]: Yes, it is set to 2

File: VMBackup/main/handle.py (hunk 7 above)
Comment (arisettisanjana) [reply VMBackup/main/handle.py:80]: To get/fetch the evnets folder mainly

File: VMBackup/main/Utils/EventLoggerUtil.py (hunk 9 above)
Comment (arisettisanjana) [reply VMBackup/main/Utils/EventLoggerUtil.py:76]: ya that is set to 2.
This just takes care of the severity of the message we are logging.

Comment (ankitha-p) [reply VMBackup/main/Utils/EventLoggerUtil.py:76]: Line 76 assumes 0. Can you update it to 2?

File: VMBackup/main/Utils/EventLoggerUtil.py [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+        self.current_message_len = 0
+        self.event_logging_enabled = False
+        self.event_logging_error_count = 0
+        self.events_folder = event_directory
+        self.event_logging_enabled = bool(self.events_folder)
+        self.filehelper = FileHelpers()
+
+        if self.event_logging_enabled:
+            self.extension_version = os.path.basename(os.getcwd())
+            self.operation_id = ''
```
Comment (ankitha-p) [inline VMBackup/main/Utils/EventLoggerUtil.py:37]: To keep consistent with Windows, lets make it an emtpy guid which will have all Zeros.

File: VMBackup/main/handle.py (hunk 11 above)
Comment (arisettisanjana) [reply VMBackup/main/handle.py:327]: discussed offline

File: VMBackup/main/handle.py (hunk 12 above)
Comment (arisettisanjana) [reply VMBackup/main/handle.py:618]: discussed offline.
as we are creating a subprocess they are having different memory space

File: VMBackup/main/Utils/EventLoggerUtil.py (hunk 9 above)
Comment (arisettisanjana) [reply VMBackup/main/Utils/EventLoggerUtil.py:76]: discussed offline

**Comment by iamashwani-who** [comment]
Can we rename the title of PR to "Adding Guest Agent based logging to VmSnapshotLinux extension".

File: VMBackup/main/Utils/EventLoggerUtil.py [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+        self.events_folder = event_directory
+        self.event_logging_enabled = bool(self.events_folder)
+        self.filehelper = FileHelpers()
+
+        if self.event_logging_enabled:
+            self.extension_version = os.path.basename(os.getcwd())
+            self.operation_id = uuid.UUID(int=0)
+            self.log_severity_level = severity_level
+            print("Information: EventLogging severity level setting is ",self.log_severity_level)
+			# creating a temp directory
```
Comment (iamashwani-who) [inline VMBackup/main/Utils/EventLoggerUtil.py:41]: nit:indentation of the comment.

**Comment by iamashwani-who** [comment]
Can we check if there is a way to turn off the guest agent based logging in case required?

File: VMBackup/main/Utils/EventLoggerUtil.py [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+        self.event_logging_error_count = 0
+        self.events_folder = event_directory
+        self.event_logging_enabled = bool(self.events_folder)
+        self.filehelper = FileHelpers()
+
+        if self.event_logging_enabled:
+            self.extension_version = os.path.basename(os.getcwd())
+            self.operation_id = uuid.UUID(int=0)
+            self.log_severity_level = severity_level
+            print("Information: EventLogging severity level setting is ",self.log_severity_level)
```
Comment (ankitha-p) [inline VMBackup/main/Utils/EventLoggerUtil.py:40]: let's add a try-catch block for the if logic

File: VMBackup/main/Utils/EventLoggerUtil.py [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+                self.current_message_len = len(message)
+            else:
+                self.current_message += message
+                self.current_message_len += len(message)
+        except Exception as ex:
+            print("Warning: Error adding extension event to queue. Exception: " + str(ex))
+
+    def begin_event_queue_polling(self):
+        print("Event polling is starting...")
+        self.event_processing_task = threading.Thread(target=self._event_processing_loop)
```
Comment (ankitha-p) [inline VMBackup/main/Utils/EventLoggerUtil.py:129]: add try:except

File: VMBackup/main/Utils/EventLoggerUtil.py [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+        while not self.event_processing_signal.wait(self.event_processing_interval):
+            try:
+                self._process_events()
+            except Exception as ex:
+                print("Warning: Event processing has failed. Exception: " + str(ex))
+        print("Information: Exiting function polling...")
+
+    def _process_events(self):
+        print("in process event")
+        if self.space_available_in_event_directory == 0:
```
Comment (ankitha-p) [inline VMBackup/main/Utils/EventLoggerUtil.py:143]: add try-except to be safe

File: VMBackup/main/Utils/EventLoggerUtil.py (hunk 18 above)
Comment (arisettisanjana) [reply VMBackup/main/Utils/EventLoggerUtil.py:40]: Any exception here would be covered when GetInstance Method is called, so avoided it.

File: VMBackup/main/Utils/EventLoggerUtil.py [hunk 21]
```diff
@@ -0,0 +1,264 @@
...
+            err_msg,
+            lambda: file.write(json_data)
+        )
+
+    def _send_event_file_to_event_directory(self, file_path, events_folder, space_available_in_event_directory):
+        file_info = os.stat(file_path)
+        file_size = file_info.st_size
+
+        if space_available_in_event_directory - file_size >= 0:
+            new_path_for_event_file = os.path.join(events_folder, os.path.basename(file_path))
```
Comment (ankitha-p) [inline VMBackup/main/Utils/EventLoggerUtil.py:220]: please use try catch as appropriate and move this to the LogHelper file.

File: VMBackup/main/Utils/StringHelper.py [hunk 22]
```diff
@@ -0,0 +1,9 @@
+import datetime
+
+class StringHelper:
+    def resolve_string(self,severity_level, message):
+        msg_body = datetime.datetime.utcnow().isoformat() + "\t" + "[" + severity_level + "]:\t"
+        if message and message.strip():
+            msg_body += message + " "
+        msg_body += "\n"
```
Comment (ankitha-p) [inline VMBackup/main/Utils/StringHelper.py:8]: please add try, except block here.

File: VMBackup/main/Utils/HandlerUtil.py [hunk 23]
```diff
@@ -164,10 +174,14 @@ def log_with_no_try_except(self, message, level='Info'):
...
                 if self.logging_file is not None:
                     self.log_py3(message)
+                    if self.eventlogger != None:
+                        self.eventlogger.trace_message(level, message)
                 else:
                     pass
             else:
                 self._log(self._get_log_prefix() + message)
+                if self.eventlogger != None:
+                    self.eventlogger.trace_message(level, message)
```
Comment (ankitha-p) [inline VMBackup/main/Utils/HandlerUtil.py:184]: better to put a try, except here to be safe (like we did in compositeLogger for windows) so that should any unexpected error occurs in trace_message it doesn't block the main run. in except we can just pass.

**Review by ankitha-p — APPROVED** [review]

**Review by iamashwani-who — APPROVED** [review]

**Review by rakrashba — APPROVED** [review]
